medicines/api/src/substance.rs

```diff
@@ -28,78 +32,83 @@ impl Substance {
 pub async fn get_substances_starting_with_letter(
     client: &impl Search,
     letter: char,
-) -> reqwest::Result<Vec<Substance>> {
+) -> anyhow::Result<Vec<Substance>> {
     let upper_letter = letter.to_ascii_uppercase();
 
     let azure_result = client
         .filter_by_collection_field("facets", &upper_letter.to_string())
         .await?;
 
-    Ok(format_search_results(azure_result, upper_letter))
+    Ok(format_search_results(azure_result, upper_letter)?)
 }
 
-fn format_search_results(results: IndexResults, letter: char) -> Vec<Substance> {
-    let mut substances: BTreeMap<&str, BTreeMap<&str, Vec<Document>>> = BTreeMap::new();
+fn format_search_results(
+    results: IndexResults,
+    letter: char,
+) -> Result<Vec<Substance>, DocTypeParseError> {
+    let mut substances: BTreeMap<String, BTreeMap<String, Vec<Document>>> = BTreeMap::new();
 
     let letter_string = letter.to_string();
 
-    results
+    let documents = results
         .search_results
-        .iter()
+        .into_iter()
         .filter(|result| result.facets.iter().any(|s| s == &letter_string))
         .filter_map(|result| {
-            let substance_name = result
-                .substance_name
-                .iter()
-                .find(|&s| s.starts_with(letter))?
-                .as_str();
-
-            let product_name = result.product_name.as_ref()?.as_str();
-
-            let doc: Document = result.clone().into();
-
-            Some((substance_name, product_name, doc))
-        })
-        .for_each(
-            |(substance, product, doc)| match substances.get_mut(substance) {
+            let doc: Document = result.into();
+
+            let substance = doc
+                .substances()
+                .find(|s| s.starts_with(letter))?
+                .to_string();
+
+            let product = doc.product_name()?.to_string();
+
+            Some((substance, product, doc))
+        });
+
+    for (substance, product, doc) in documents {
```

Comment:
can this be broken out into another method? this `format_search_results` is 52 lines long